### DISCUSSION TITLE
The lifetime of C++ std::function converted from python function

### DISCUSSION THREAD

**Discussion by metab0t** [wjakob/nanobind#528] [Q&A] [answered]
In C++, we can convert a Python function to std::function.
```cpp
class CClass
{
    std::function<int(int)> f;
    void set_f(const std::function<int(int)>& f_)
    {
        f = f_;
    }
};

NB_MODULE(gurobi_model_ext, m)
{
    nb::class_<CClass>(m, "CClass").def(nb::init<>()).def(set_f, &CClass::set_f);
}
```

```python
class PClass:
    def __init__(self):
        self.x = 1
    def g(y):
        return x + y

P = PClass()
C = CClass()
C.set_f(P.g)
```

I am not sure whether nanobind will ensure the liveness of `P` as long as `C` is alive?

**Comment by wjakob** [comment, marked as answer]
Nanobind will keep the object alive to satisfy the dependency, if that's what you are asking.

**Reply by metab0t** [reply]
Thank you! Now I am confident to pass Python function as callback to C++ side.

I am currently using nanobind to write a new modeling interface for mathematical optimization called [PyOptInterface](https://github.com/metab0t/PyOptInterface) and the efficient C++/Python communication brought by nanobind has shown very promising performance.
